Add command to trigger a snapshot in primary master journal system

LuQQiu opened the pull request:
Add a SnapshotCommand to trigger a snapshot/checkpoint in primary master journal system.
This command is added for debugging or avoid primary master log from going unbounded.

LuQQiu commented:
jenkins, test this please

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/Alluxio-Pull-Request-Builder/3237/
Test FAILed.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/Alluxio-Pull-Request-Builder/3238/
Test FAILed.

LuQQiu commented:
Jenkins, test this please

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/Alluxio-Pull-Request-Builder/3239/
Test FAILed.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/Alluxio-Pull-Request-Builder/3240/
Test FAILed.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/Alluxio-Pull-Request-Builder/3243/
Test FAILed.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/Alluxio-Pull-Request-Builder/3244/
Test PASSed.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/Alluxio-Pull-Request-Builder/3246/
Test PASSed.

aaudiber commented on core/server/common/src/main/java/alluxio/master/journal/raft/RaftJournalSystem.java:
```diff
@@ -327,6 +328,63 @@ public synchronized void losePrimacy() {
+        return builder.setTriggered(false)
+            .setMessage("Interrupted while submitting entry to trigger snapshot").build();
+      }
+      if (!mStateMachine.isSnapshotting()) {
```
could the snapshot have already finished?

LuQQiu replied:
I tested on the alluxio cluster with fsmeta operations, the snapshot always fall into this if block. In this way, we are not exactly sure if `statemachine` snapshots. I will come up with other ways to make sure snapshot happens. You are right, the snapshot could have already finished.

LuQQiu replied:
Add a `mLastSnapshotStartTime` to trigger when the last snapshot start. If it is bigger than the `RaftJournalSystem.checkpoint()` start time, we successfully trigger the snapshot.

aaudiber commented on core/server/common/src/main/java/alluxio/master/journal/raft/RaftJournalSystem.java:
```diff
@@ -327,6 +328,63 @@ public synchronized void losePrimacy() {
+      CommonUtils.waitFor("snapshotting to finish", () -> !stateMachine.isSnapshotting(),
+          WaitForOptions.defaults().setTimeoutMs(20 * Constants.MINUTE_MS));
+    } catch (InterruptedException e) {
+      Thread.currentThread().interrupt();
```
when would we hit this code?

LuQQiu replied:
The `CommonUtils.waitFor()` method requires handling of the `InterruptedException`, we may not be able to hit this code.

aaudiber commented on core/server/common/src/main/java/alluxio/master/journal/raft/RaftJournalSystem.java:
```diff
@@ -327,6 +328,63 @@ public synchronized void losePrimacy() {
+      }
+      return builder.setTriggered(true).setSucceed(true).build();
+    } finally {
+      if (mPrimarySelector.getState() == PrimarySelector.State.PRIMARY) {
```
this method is synchronized with `gainPrimacy` and `losePrimacy`, so I think it's safe to always set `mSnapshotAllowed` to `false` at the end. If the primary selector switched to secondary while the snapshot was happening, we will call `losePrimacy` afterward, which will set `mSnapshotAllowed` to `true`

LuQQiu replied:
Make sense, remove the Primary selector state check

aaudiber commented on core/server/common/src/main/java/alluxio/master/journal/ufs/UfsJournal.java:
```diff
@@ -311,6 +312,19 @@ public void format() throws IOException {
+   * @return the snapshot response
+   */
+  public synchronized SnapshotPResponse checkpoint() {
+    if (mTailerThread == null) {
```
isn't this normally `null` on the primary?

LuQQiu replied:
Oh yeah, now change to directly `mMaster.writeToCheckpoint(journalWriter);` now.

aaudiber commented on core/transport/src/grpc/meta_master.proto:
```diff
@@ -112,6 +112,20 @@ message GetMetricsPResponse {
+message SnapshotPResponse {
+ optional bool triggered = 1;
+ optional bool succeed = 2;
+ optional string message = 3;
```
usually we use exceptions for this sort of thing. If the rpc fails to trigger a snapshot, we throw an exception saying why it failed to trigger. If the rpc starts a snapshot but the snapshot fails, we throw an exception saying why the snapshot failed. If we don't throw anything, that means it succeeded.

LuQQiu replied:
Yeah, will use exceptions instead of message here to indication the rpc fails.
But Snapshot has one more circumstances other than snapshotted or failed: the snapshot is not triggered. 

Is it OK that we only responsible for giving the possibility of taking a snapshot, but not guarantee snapshotted in the Snapshot CLI? Giving console message like: successfully send the snapshot signal.

aaudiber replied:
I think we can throw an exception with a relevant message if we fail to checkpoint because the journal doesn't want to be checkpointed, e.g. Copycat decides not to snapshot, or in the UFS journal no entries have been written since the last checkpoint. I don't see a need for the client side to differentiate between "Failed because the journal refused" and "Failed for some other reason". Either way I think we'll just forward the error message to the client

LuQQiu replied:
Make sense. Now I will also throw exceptions when not be able to trigger snapshot. Only return results when snapshot is triggered and succeed.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/Alluxio-Pull-Request-Builder/3261/
Test FAILed.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/Alluxio-Pull-Request-Builder/3262/
Test PASSed.

LuQQiu commented on core/client/fs/src/main/java/alluxio/client/meta/MetaMasterClient.java:
```diff
@@ -56,4 +57,11 @@
+   *
+   * @return the snapshot response
+   */
+  SnapshotPResponse snapshot() throws IOException;
```
Change the name from snapshot to checkpoint. Thanks

LuQQiu commented on core/client/fs/src/main/java/alluxio/client/meta/MetaMasterClient.java:
```diff
@@ -56,4 +57,11 @@
+   *
+   * @return the snapshot response
+   */
+  SnapshotPResponse snapshot() throws IOException;
```
Now i only return the hostname of the master that trigger the checkpoint.

LuQQiu commented on core/server/common/src/main/java/alluxio/master/journal/raft/RaftJournalSystem.java:
```diff
@@ -327,6 +328,63 @@ public synchronized void losePrimacy() {
+    mSnapshotAllowed.set(true);
+    try {
+      CopycatClient client = createClient();
+      long sequence = ThreadLocalRandom.current().nextLong(Long.MIN_VALUE, 0);
```
Changed to use `JournalEntry.getDefaultInstance()`, thanks

LuQQiu commented on core/server/common/src/main/java/alluxio/master/journal/raft/RaftJournalSystem.java:
```diff
@@ -327,6 +328,63 @@ public synchronized void losePrimacy() {
+      try {
+        future.get(5, TimeUnit.SECONDS);
+      } catch (TimeoutException | ExecutionException e) {
+        LOG.info("Exception submitting entry: {}", e.toString());
```
Also throw exception here.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/Alluxio-Pull-Request-Builder/3276/
Test PASSed.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/Alluxio-Pull-Request-Builder/3283/
Test PASSed.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/Alluxio-Pull-Request-Builder/3284/
Test PASSed.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/Alluxio-Pull-Request-Builder/3285/
Test PASSed.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/Alluxio-Pull-Request-Builder/3286/
Test PASSed.

LuQQiu commented:
@aaudiber Addressed comments and tested in Alluxio cluster with embedded journal and ufs journal. Now is working on adding tests. PTAL, thanks

aaudiber commented on core/server/common/src/main/java/alluxio/master/journal/noop/NoopJournalSystem.java:
```diff
@@ -57,4 +57,9 @@ public void start() {}
+
+  @Override
+  public void checkpoint() {
+    throw new UnsupportedOperationException("Checkpoint is not supported in NoopJournalSystem");
```
more consistent to make this a no-op like other methods

LuQQiu replied:
Changed, thanks

aaudiber commented on core/server/common/src/main/java/alluxio/master/journal/raft/RaftJournalSystem.java:
```diff
@@ -327,6 +334,58 @@ public synchronized void losePrimacy() {
+      try {
+        future.get(1, TimeUnit.MINUTES);
+      } catch (TimeoutException | ExecutionException e) {
+        LOG.info("Exception submitting entry to trigger snapshot: {}", e.toString());
```
can this be LOG.warn?

LuQQiu replied:
Changed, thanks

aaudiber commented on core/server/common/src/main/java/alluxio/master/journal/raft/RaftJournalSystem.java:
```diff
@@ -327,6 +334,58 @@ public synchronized void losePrimacy() {
+        LOG.info("Exception submitting entry to trigger snapshot: {}", e.toString());
+        throw new IOException("Exception submitting entry to trigger snapshot", e);
+      } catch (InterruptedException e) {
+        LOG.info("Interrupted when submitting entry to trigger snapshot: {}", e.toString());
```
ditto

LuQQiu replied:
Changed, thanks

aaudiber commented on core/server/common/src/main/java/alluxio/master/journal/raft/RaftJournalSystem.java:
```diff
@@ -327,6 +334,58 @@ public synchronized void losePrimacy() {
+      if (mStateMachine.isSnapshotting()) {
+        waitForSnapshotting(mStateMachine);
+      }
+      LOG.info("Snapshotted in raft journal system");
```
I think the snapshot implementation will already log about this, so we don't need a second log here

LuQQiu replied:
Yeah, we don't need a second log here, removed, thanks

aaudiber commented on core/server/common/src/main/java/alluxio/master/journal/raft/RaftJournalSystem.java:
```diff
@@ -327,6 +334,58 @@ public synchronized void losePrimacy() {
+  @Override
+  public synchronized void checkpoint() throws IOException {
+    mSnapshotAllowed.set(true);
+    CopycatClient client = createAndConnectClient();
```
this should be inside the `try` - otherwise if we throw an exception, we won't set `mSnapshotAllowed` back to `false`

LuQQiu replied:
Move inside the try finally. Thanks

aaudiber commented on core/server/common/src/main/java/alluxio/master/journal/raft/RaftJournalSystem.java:
```diff
@@ -327,6 +334,58 @@ public synchronized void losePrimacy() {
+        Thread.currentThread().interrupt();
+        throw new CancelledException("Interrupted when submitting entry to trigger snapshot", e);
+      }
+      if (mStateMachine.getLastSnapshotStartTime() < start) {
```
what if the snapshot is about to start, but just hasn't started yet? Perhaps we could retry for some period before giving up

LuQQiu replied:
Add the retry check start time, thanks

aaudiber commented on core/server/common/src/main/java/alluxio/master/journal/raft/RaftJournalSystem.java:
```diff
@@ -327,6 +334,58 @@ public synchronized void losePrimacy() {
+  private void waitForSnapshotting(JournalStateMachine stateMachine) throws IOException {
+    try {
+      CommonUtils.waitFor("snapshotting to finish", () -> !stateMachine.isSnapshotting(),
+          WaitForOptions.defaults().setTimeoutMs(20 * Constants.MINUTE_MS));
```
unrelated to this PR, but can we make this configurable and increase the default to 2 hours? With 1 billion files this might time out due to the large scale

LuQQiu replied:
Add a property key to control how long we wait for the snapshotting to finish.

aaudiber commented on core/server/common/src/main/java/alluxio/master/journal/ufs/UfsJournal.java:
```diff
@@ -311,6 +312,31 @@ public void format() throws IOException {
+    long nextSequenceNumber = getNextSequenceNumberToWrite();
+    if (nextSequenceNumber - getNextSequenceNumberToCheckpoint() < ServerConfiguration.getLong(
+        PropertyKey.MASTER_JOURNAL_CHECKPOINT_PERIOD_ENTRIES)) {
+      LOG.info("{}: No enough journal entries to checkpoint.",
```
can we force a checkpoint anyway in this situation? The user is indicating that they want a checkpoint, so we should do our best to create one

LuQQiu replied:
Now only when the `nextSequenceNumber == getNextSequenceNumberToCheckpoint()` we don't trigger snapshot.

aaudiber commented on core/server/common/src/main/java/alluxio/master/journal/ufs/UfsJournalSystem.java:
```diff
@@ -154,4 +154,11 @@ public void format() throws IOException {
+
+  @Override
+  public void checkpoint() throws IOException {
+    for (UfsJournal  journal : mJournals.values()) {
```
nit: extra space

LuQQiu replied:
Fixed, thanks

aaudiber commented on shell/src/main/java/alluxio/cli/fsadmin/command/CheckpointCommand.java:
```diff
@@ -0,0 +1,75 @@
+    timer.schedule(new TimerTask() {
+        @Override
+        public void run() {
+          mPrintStream.println("Checkpointing may take a while.");
```
this is pretty verbose - in other places we've logged a single "." every few seconds without adding newlines, see https://github.com/Alluxio/alluxio/blob/master/job/client/src/main/java/alluxio/client/job/JobGrpcClientUtils.java#L97

LuQQiu replied:
Used the `...`  interpreting waiting

aaudiber commented on shell/src/main/java/alluxio/cli/fsadmin/command/CheckpointCommand.java:
```diff
@@ -0,0 +1,75 @@
+    }, Constants.MINUTE_MS);
+    String masterHostname = mMetaClient.checkpoint();
+    timer.cancel();
+    mPrintStream.println("Successfully checkpointed in " + masterHostname);
```
`Successfully took a checkpoint on master %s`

LuQQiu replied:
Changed, thanks

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/Alluxio-Pull-Request-Builder/3306/
Test PASSed.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/Alluxio-Pull-Request-Builder/3309/
Test PASSed.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/Alluxio-Pull-Request-Builder/3326/
Test PASSed.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/Alluxio-Pull-Request-Builder/3328/
Test PASSed.

LuQQiu commented:
@aaudiber Look into the snapshot requirements, the main requirement that always fails to achieve is Copycat need to have new segments (I think it is a file contains many logs). The segment size is controlled by `alluxio.master.journal.log.size.bytes.max` which by default is `10MB`. When I set it to `1kb`, the snapshot will be triggered and succeed definitely.
Addressed the comments, add the triggered embedded journal snapshot integration test and wait for snapshot to start logic. PTAL, thanks

aaudiber commented on core/server/common/src/main/java/alluxio/master/journal/raft/RaftJournalSystem.java:
```diff
@@ -327,6 +334,79 @@ public synchronized void losePrimacy() {
+      Thread.currentThread().interrupt();
+      throw new CancelledException("Interrupted when waiting for snapshotting to start", e);
+    } catch (TimeoutException e) {
+      throw new IOException("Do not fulfill Copycat snapshot requirements. "
```
this message will likely be propagated to users, who might not know about snapshots or copycat. Can we say something more general like "A recent checkpoint already exists, not creating a new one"?

LuQQiu replied:
Yeah, changed. Now is more user-friendly.

aaudiber commented on core/server/common/src/main/java/alluxio/master/journal/raft/RaftJournalSystem.java:
```diff
@@ -327,6 +334,79 @@ public synchronized void losePrimacy() {
+        throw new CancelledException("Interrupted when submitting entry to trigger snapshot", e);
+      }
+      waitForSnapshotStart(mStateMachine, start);
+      if (mStateMachine.isSnapshotting()) {
```
unnecessary check

LuQQiu replied:
Removed the check

aaudiber commented on core/server/common/src/main/java/alluxio/master/journal/ufs/UfsJournal.java:
```diff
@@ -311,6 +312,29 @@ public void format() throws IOException {
   }
 
+  /**
+   * Triggers a checkpoint in this ufs journal.
```
"Triggers" suggests asynchronous, I think "Creates" is a better word here

LuQQiu replied:
Changed the `triggers` to `creates` to sound more synchronous

aaudiber commented on core/server/common/src/main/java/alluxio/master/journal/ufs/UfsJournal.java:
```diff
@@ -311,6 +312,29 @@ public void format() throws IOException {
+  public synchronized void checkpoint() throws IOException {
+    long nextSequenceNumber = getNextSequenceNumberToWrite();
+    if (nextSequenceNumber == getNextSequenceNumberToCheckpoint()) {
+      LOG.info("{}: No enough journal entries to create checkpoint.",
```
We can be more specific, saying that no entries have been written since the last checkpoint

LuQQiu replied:
Changed, more clear. thanks

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/Alluxio-Pull-Request-Builder/3354/
Test PASSed.

LuQQiu commented:
@aaudiber Thanks for correcting the documentation, Addressed the comments, PTAL, thanks

aaudiber commented:
alluxio-bot, merge this please
